Cython/Compiler/ExprNodes.py

```diff
@@ -14271,8 +14271,34 @@ def analyse_type_annotation(self, env, assigned_value=None):
         with env.new_c_type_context(in_c_type_context=explicit_ctype):
             arg_type = annotation.analyse_as_type(env)
 
+
         if arg_type is None:
-            warning(annotation.pos, "Unknown type declaration in annotation, ignoring")
+            if isinstance(annotation, NameNode):
+                # Validate annotation in form `var: type`
+                if not env.lookup(annotation.name):
+                    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
+            elif isinstance(annotation, AttributeNode):
+                # Validate annotation in form `var: module.type`
+                if not env.lookup(annotation.obj.name):
```

Comment:
I'm not sure this is necessarily referring to a module. And even if it was, looking up the attribute name in the current scope won't find the name in that module. There's `typing` module support now, so we might want to special case that. And the `.entry` of the attribute's object might also give hints.

Reply:
I am pretty new to Cython code-base so maybe here I need a help. This test:

```python
if not env.lookup(annotation.obj.name):
    # `module` is undeclared
    warning(annotation.pos, ...)
```

is checking whether for the annotation expression `A.B`  is `A` name known/defined. If not, it is raising the warning. Is this incorrect? It is pretty the same as previous test:
```
if not env.lookup(annotation.name):
    warning(annotation.pos, "Unknown type declaration '%s' in annotation, ignoring" % annotation.name, level=1)
```

Reply:
Not sure if this is quite the point that scoder was making, but

```
annotation.obj.name
```

could fail. `annotation.obj` isn't necessarily a name node. Some (non-exhaustive) examples:

```
mod.a.b  # I think obj will be mod.a
func().b  # obj will be func()  - I can't imagine many times this'll be a useful annotation for Cython, but it shouldn't crash!
(1 + x).b  # again, unlikely
```

There's no need to produce a sensible message for all of these, but they shouldn't crash

Reply:
Now I understand. I have limited the validation  only when `annotation.obj.is_name` is true. So for simplicity we won't manage all nasty corner cases. Let me know if you a fine with this solution.